src/zarr/store/remote.py

```diff
@@ -11,90 +14,141 @@
     from fsspec.asyn import AsyncFileSystem
     from upath import UPath
 
+    from zarr.buffer import Buffer
+    from zarr.common import BytesLike
+
 
 class RemoteStore(Store):
+    # based on FSSpec
     supports_writes: bool = True
     supports_partial_writes: bool = False
     supports_listing: bool = True
 
-    root: UPath
+    _fs: AsyncFileSystem
+    exceptions = tuple[type[Exception], ...]
 
     def __init__(
-        self, url: UPath | str, *, mode: OpenMode = "r", **storage_options: dict[str, Any]
+        self,
+        url: UPath | str,
+        mode: OpenMode = "r",
+        allowed_exceptions: tuple[type[Exception], ...] = (
+            FileNotFoundError,
+            IsADirectoryError,
+            NotADirectoryError,
+        ),
+        **storage_options: Any,
     ):
-        import fsspec
-        from upath import UPath
+        """
+        Parameters
+        ----------
+        url: root of the datastore. In fsspec notation, this is usually like "protocol://path/to".
+            Can also be a upath.UPath instance/
+        allowed_exceptions: when fetching data, these cases will be deemed to correspond to missinf
+            keys, rather than some other IO failure
+        storage_options: passed on to fsspec to make the filesystem instance. If url is a UPath,
+            this must not be used.
+        """
 
         super().__init__(mode=mode)
 
         if isinstance(url, str):
-            self.root = UPath(url, **storage_options)
+            self._fs, self.path = fsspec.url_to_fs(url, **storage_options)
+        elif hasattr(url, "protocol") and hasattr(url, "fs"):
+            # is UPath-like - but without importing
+            if storage_options:
+                raise ValueError(
+                    "If constructed with a UPath object, no additional "
+                    "storage_options are allowed"
+                )
+            self.path = url.path
+            self._fs = url._fs
         else:
-            assert (
-                len(storage_options) == 0
-            ), "If constructed with a UPath object, no additional storage_options are allowed."
-            self.root = url.rstrip("/")
-
+            raise ValueError("URL not understood, %s", url)
+        self.exceptions = allowed_exceptions
```

Comment:
@dstansby  - can you take a look here?
```
src/zarr/store/remote.py:67: error: Incompatible types in assignment (expression has type "tuple[type[Exception], ...]", variable has type "type[tuple[Any, ...]]")  [assignment]
```

Reply:
There was a typo in the definition of `exceptions` https://github.com/zarr-developers/zarr-python/pull/1785/commits/5315fda35c77bfa2cf512c7ba1cd1e8ec35256fe#diff-22c6cf4de485dcc3037bd442d219d8d1c0c56ac62ae3a40b337a0ca088c2fc25R29